File: examples/02-reliability/python/async_activity.py
from datetime import timedelta

from temporalio import workflow

# `extract`, `transform`, `load` are module-level @activity.defn functions
# (omitted here for clarity). For activities defined as class methods instead,
# use workflow.start_activity_method / execute_activity_method.


# In Python the Workflow runs on asyncio. start_activity() returns an awaitable
# handle right away; await it to get the result. Kicking off two extracts before
# awaiting either lets them run concurrently — the Async.function equivalent.
@workflow.defn
class AsyncWorkflow:
    @workflow.run
    async def run(self, batch_date: str) -> None:
        opts = dict(start_to_close_timeout=timedelta(minutes=5))

        raw = workflow.start_activity(extract, batch_date, **opts)
        audit = workflow.start_activity(extract, f"{batch_date}-audit", **opts)

        clean_uri = await workflow.execute_activity(transform, await raw, **opts)
        clean_audit_uri = await workflow.execute_activity(transform, await audit, **opts)

        await workflow.execute_activity(load, clean_uri, **opts)
        await workflow.execute_activity(load, clean_audit_uri, **opts)

File: examples/02-reliability/go/async_activity.go
package reliability

import (
	"time"

	"go.temporal.io/sdk/workflow"
)

// In Go, ExecuteActivity returns a Future immediately; .Get(ctx, &out) blocks
// for the result. Starting two extracts before getting either lets them run
// concurrently — the equivalent of Java's Async.function.
func AsyncWorkflow(ctx workflow.Context, batchDate string) error {
	ctx = workflow.WithActivityOptions(ctx, workflow.ActivityOptions{
		StartToCloseTimeout: 5 * time.Minute,
	})

	rawF := workflow.ExecuteActivity(ctx, "Extract", batchDate)
	auditF := workflow.ExecuteActivity(ctx, "Extract", batchDate+"-audit")

	var rawURI, auditURI string
	_ = rawF.Get(ctx, &rawURI)
	_ = auditF.Get(ctx, &auditURI)

	var cleanURI, cleanAuditURI string
	_ = workflow.ExecuteActivity(ctx, "Transform", rawURI).Get(ctx, &cleanURI)
	_ = workflow.ExecuteActivity(ctx, "Transform", auditURI).Get(ctx, &cleanAuditURI)

	_ = workflow.ExecuteActivity(ctx, "Load", cleanURI).Get(ctx, nil)
	return workflow.ExecuteActivity(ctx, "Load", cleanAuditURI).Get(ctx, nil)
}
